File: kirun-js/src/engine/function/sytem/math/Hypotenuse.ts
import { Schema } from '../../../json/schema/Schema';
import { SchemaType } from '../../../json/schema/type/SchemaType';
import { TypeUtil } from '../../../json/schema/type/TypeUtil';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { PrimitiveUtil } from '../../../util/primitive/PrimitiveUtil';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'value';

export class Hypotenuse extends AbstractFunction {
    private static readonly SIGNATURE: FunctionSignature = new FunctionSignature()
        .setName('Hypotenuse')
        .setNamespace(Namespaces.MATH)
        .setParameters(
            new Map([
                [
                    VALUE,
                    new Parameter()
                        .setParameterName(VALUE)
                        .setSchema(Schema.ofNumber(VALUE))
                        .setVariableArgument(true),
                ],
            ]),
        )
        .setEvents(
            new Map([
                Event.outputEventMapEntry(
                    new Map([
                        [
                            VALUE,
                            new Schema().setType(TypeUtil.of(SchemaType.DOUBLE)).setName(VALUE),
                        ],
                    ]),
                ) as [string, Event],
            ]),
        );

    public constructor() {
        super();
    }

    public getSignature(): FunctionSignature {
        return Hypotenuse.SIGNATURE;
    }

    protected internalExecute(context: FunctionExecutionParameters): FunctionOutput {
        let nums: number[] = context.getArguments().get(VALUE);

        return new FunctionOutput([
            EventResult.outputOf(
                new Map([[VALUE, Math.sqrt(nums.reduce((a, c) => (a += c * c), 0))]]),
            ),
        ]);
    }
}

File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/math/Hypotenuse.java
package com.fincity.nocode.kirun.engine.function.system.math;

import static com.fincity.nocode.kirun.engine.namespaces.Namespaces.MATH;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.function.AbstractFunction;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.Event;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.FunctionSignature;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.runtime.FunctionExecutionParameters;
import com.fincity.nocode.kirun.engine.util.string.StringFormatter;
import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

public class Hypotenuse extends AbstractFunction {

	private static final String VALUE = "value";

	private static final FunctionSignature SIGNATURE = new FunctionSignature().setName("Hypotenuse").setNamespace(MATH)
			.setParameters(Map.of(VALUE,
					new Parameter().setParameterName(VALUE).setSchema(Schema.ofNumber(VALUE))
							.setVariableArgument(true)))
			.setEvents(Map.ofEntries(Event.outputEventMapEntry(Map.of(VALUE, Schema.ofNumber(VALUE)))));

	@Override
	public FunctionSignature getSignature() {
		return SIGNATURE;
	}

	@Override
	protected FunctionOutput internalExecute(FunctionExecutionParameters context) {

		JsonElement jsonElement = context.getArguments().get(VALUE);

		if (!jsonElement.isJsonArray()) {
			throw new KIRuntimeException(StringFormatter.format("Expected an array but found $", jsonElement));
		}

		if (jsonElement.getAsJsonArray().isEmpty()) {

			return new FunctionOutput(List.of(EventResult.outputOf(Map.of(VALUE, new JsonPrimitive(0)))));
		}

		Iterator<JsonElement> iterator = jsonElement.getAsJsonArray().iterator();

		double sumOfSqaures = 0;

		while (iterator.hasNext()) {
			double value = iterator.next().getAsDouble();
			sumOfSqaures += value * value;
		}

		JsonPrimitive element = new JsonPrimitive(Math.sqrt(sumOfSqaures));

		return new FunctionOutput(List.of(EventResult.outputOf(Map.of(VALUE, element))));
	}
}
